File: examples/onnx_inference_example.js
/**
 * LaMa ONNX Inference Example - JavaScript/Node.js
 *
 * This script demonstrates the correct preprocessing and postprocessing for LaMa ONNX model.
 *
 * CRITICAL INFORMATION:
 * - Input images and masks must be normalized to [0, 1] range
 * - Output is already in [0, 255] range (DO NOT multiply by 255 again!)
 * - All inputs must be in CHW (Channel, Height, Width) format with batch dimension
 *
 * Installation:
 *   npm install onnxruntime-node sharp
 *
 * Usage:
 *   node onnx_inference_example.js model.onnx input.jpg mask.png output.jpg
 *
 * For more details, see the README's "ONNX Preprocessing & Postprocessing Guide" section.
 */

const ort = require('onnxruntime-node');
const sharp = require('sharp');
const fs = require('fs').promises;

/**
 * Convert image buffer to CHW format tensor normalized to [0, 1]
 *
 * @param {Buffer} imageBuffer - Image buffer from sharp
 * @param {number} width - Image width
 * @param {number} height - Image height
 * @param {number} channels - Number of channels (3 for RGB, 1 for grayscale)
 * @returns {Float32Array} - Flattened CHW tensor normalized to [0, 1]
 */
function preprocessImage(imageBuffer, width, height, channels) {
    // Input buffer is in HWC format (height, width, channels)
    // We need to convert to CHW format and normalize to [0, 1]

    const size = width * height * channels;
    const tensor = new Float32Array(size);

    // Convert HWC to CHW and normalize to [0, 1]
    for (let c = 0; c < channels; c++) {
        for (let h = 0; h < height; h++) {
            for (let w = 0; w < width; w++) {
                const hwcIndex = (h * width + w) * channels + c;
                const chwIndex = c * height * width + h * width + w;

                // CRITICAL: Divide by 255 to normalize to [0, 1]
                tensor[chwIndex] = imageBuffer[hwcIndex] / 255.0;
            }
        }
    }

    return tensor;
}

/**
 * Pad image dimensions to be divisible by mod
 *
 * @param {number} dimension - Original dimension
 * @param {number} mod - Modulo value (typically 8)
 * @returns {number} - Padded dimension
 */
function padToModulo(dimension, mod = 8) {
    return Math.ceil(dimension / mod) * mod;
}

/**
 * Convert CHW output tensor to HWC image buffer
 *
 * @param {Float32Array} output - Model output in CHW format
 * @param {number} width - Image width
 * @param {number} height - Image height
 * @param {number} channels - Number of channels (3 for RGB)
 * @returns {Uint8Array} - Image buffer in HWC format
 */
function postprocessOutput(output, width, height, channels) {
    const size = width * height * channels;
    const imageBuffer = new Uint8Array(size);

    // Convert CHW to HWC
    for (let c = 0; c < channels; c++) {
        for (let h = 0; h < height; h++) {
            for (let w = 0; w < width; w++) {
                const chwIndex = c * height * width + h * width + w;
                const hwcIndex = (h * width + w) * channels + c;

                // CRITICAL: Output is already in [0, 255] range!
                // DO NOT multiply by 255 again!
                // Just clamp and convert to uint8
                const value = Math.max(0, Math.min(255, output[chwIndex]));
                imageBuffer[hwcIndex] = value;
            }
        }
    }

    return imageBuffer;
}

/**
 * Run LaMa ONNX inference
 *
 * @param {string} modelPath - Path to ONNX model
 * @param {string} imagePath - Path to input image
 * @param {string} maskPath - Path to mask image
 * @param {string} outputPath - Path to save output
 */
async function runInference(modelPath, imagePath, maskPath, outputPath) {
    console.log(`Loading ONNX model from ${modelPath}...`);
    const session = await ort.InferenceSession.create(modelPath);

    console.log(`Loading and preprocessing image from ${imagePath}...`);
    const imageMetadata = await sharp(imagePath).metadata();
    const originalWidth = imageMetadata.width;
    const originalHeight = imageMetadata.height;

    // Pad dimensions to be divisible by 8
    const paddedWidth = padToModulo(originalWidth, 8);
    const paddedHeight = padToModulo(originalHeight, 8);

    console.log(`  Original size: ${originalWidth}x${originalHeight}`);
    console.log(`  Padded size: ${paddedWidth}x${paddedHeight}`);

    // Load and preprocess image
    const imageBuffer = await sharp(imagePath)
        .resize(paddedWidth, paddedHeight, { fit: 'contain', background: { r: 0, g: 0, b: 0 } })
        .raw()
        .toBuffer();

    const imageArray = preprocessImage(imageBuffer, paddedWidth, paddedHeight, 3);

    console.log(`Loading and preprocessing mask from ${maskPath}...`);
    // Load and preprocess mask (convert to grayscale)
    const maskBuffer = await sharp(maskPath)
        .resize(paddedWidth, paddedHeight, { fit: 'contain', background: { r: 0, g: 0, b: 0 } })
        .grayscale()
        .raw()
        .toBuffer();

    const maskArray = preprocessImage(maskBuffer, paddedWidth, paddedHeight, 1);

    // Binarize mask (values > 0.5 become 1.0, else 0.0)
    for (let i = 0; i < maskArray.length; i++) {
        maskArray[i] = maskArray[i] > 0.5 ? 1.0 : 0.0;
    }

    // Create input tensors with batch dimension
    const imageTensor = new ort.Tensor('float32', imageArray, [1, 3, paddedHeight, paddedWidth]);
    const maskTensor = new ort.Tensor('float32', maskArray, [1, 1, paddedHeight, paddedWidth]);

    console.log(`Running inference...`);
    console.log(`  Image tensor shape: [${imageTensor.dims.join(', ')}]`);
    console.log(`  Mask tensor shape: [${maskTensor.dims.join(', ')}]`);

    // Run inference
    const feeds = { image: imageTensor, mask: maskTensor };
    const results = await session.run(feeds);
    const output = results.output;

    console.log(`  Output tensor shape: [${output.dims.join(', ')}]`);
    console.log(`  Output range: [${Math.min(...output.data)}, ${Math.max(...output.data)}]`);

    // Postprocess output
    console.log(`Postprocessing output...`);
    const outputBuffer = postprocessOutput(
        output.data,
        paddedWidth,
        paddedHeight,
        3
    );

    // Crop to original size and save
    console.log(`Saving result to ${outputPath}...`);
    await sharp(outputBuffer, {
        raw: {
            width: paddedWidth,
            height: paddedHeight,
            channels: 3
        }
    })
    .extract({ left: 0, top: 0, width: originalWidth, height: originalHeight })
    .toFile(outputPath);

    console.log('✅ Done!');
    console.log('\nIMPORTANT NOTES:');
    console.log('  - Input was normalized to [0, 1] range');
    console.log('  - Output is already in [0, 255] range (NO additional scaling)');
    console.log('  - Result saved as uint8 image');
}

// Main entry point
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length !== 4) {
        console.error('Usage: node onnx_inference_example.js <model.onnx> <input.jpg> <mask.png> <output.jpg>');
        console.error('');
        console.error('For more information, see the README\'s "ONNX Preprocessing & Postprocessing Guide" section.');
        process.exit(1);
    }

    const [modelPath, imagePath, maskPath, outputPath] = args;

    runInference(modelPath, imagePath, maskPath, outputPath)
        .catch(err => {
            console.error('Error:', err);
            process.exit(1);
        });
}

module.exports = { runInference, preprocessImage, postprocessOutput, padToModulo };

File: examples/OnnxInferenceExample.cs
/**
 * LaMa ONNX Inference Example - C#
 *
 * This script demonstrates the correct preprocessing and postprocessing for LaMa ONNX model.
 *
 * CRITICAL INFORMATION:
 * - Input images and masks must be normalized to [0, 1] range
 * - Output is already in [0, 255] range (DO NOT multiply by 255 again!)
 * - All inputs must be in CHW (Channel, Height, Width) format with batch dimension
 *
 * Installation (NuGet packages):
 *   Microsoft.ML.OnnxRuntime
 *   SixLabors.ImageSharp
 *
 * Usage:
 *   dotnet run model.onnx input.jpg mask.png output.jpg
 *
 * For more details, see the README's "ONNX Preprocessing & Postprocessing Guide" section.
 */

using System;
using System.Linq;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;

namespace LamaOnnxInference
{
    public class OnnxInferenceExample
    {
        /// <summary>
        /// Convert image to CHW format tensor normalized to [0, 1]
        /// </summary>
        /// <param name="image">Input image</param>
        /// <returns>Float tensor in CHW format, normalized to [0, 1]</returns>
        private static float[] PreprocessImage(Image<Rgb24> image)
        {
            int width = image.Width;
            int height = image.Height;
            float[] tensor = new float[3 * height * width];

            // Convert HWC to CHW and normalize to [0, 1]
            image.ProcessPixelRows(accessor =>
            {
                for (int h = 0; h < height; h++)
                {
                    Span<Rgb24> pixelRow = accessor.GetRowSpan(h);
                    for (int w = 0; w < width; w++)
                    {
                        var pixel = pixelRow[w];

                        // CHW format: all R values, then all G values, then all B values
                        // CRITICAL: Normalize to [0, 1] by dividing by 255
                        tensor[0 * height * width + h * width + w] = pixel.R / 255.0f;
                        tensor[1 * height * width + h * width + w] = pixel.G / 255.0f;
                        tensor[2 * height * width + h * width + w] = pixel.B / 255.0f;
                    }
                }
            });

            return tensor;
        }

        /// <summary>
        /// Convert grayscale image to CHW format tensor normalized to [0, 1]
        /// </summary>
        private static float[] PreprocessMask(Image<L8> mask)
        {
            int width = mask.Width;
            int height = mask.Height;
            float[] tensor = new float[height * width];

            // Convert to CHW and normalize to [0, 1]
            mask.ProcessPixelRows(accessor =>
            {
                for (int h = 0; h < height; h++)
                {
                    Span<L8> pixelRow = accessor.GetRowSpan(h);
                    for (int w = 0; w < width; w++)
                    {
                        // CRITICAL: Normalize to [0, 1] and binarize (> 0.5 = 1.0)
                        float value = pixelRow[w].PackedValue / 255.0f;
                        tensor[h * width + w] = value > 0.5f ? 1.0f : 0.0f;
                    }
                }
            });

            return tensor;
        }

        /// <summary>
        /// Pad dimension to be divisible by mod
        /// </summary>
        private static int PadToModulo(int dimension, int mod = 8)
        {
            return (int)Math.Ceiling((double)dimension / mod) * mod;
        }

        /// <summary>
        /// Convert CHW output tensor to image
        /// </summary>
        /// <param name="output">Model output in CHW format</param>
        /// <param name="width">Image width</param>
        /// <param name="height">Image height</param>
        /// <returns>Output image</returns>
        private static Image<Rgb24> PostprocessOutput(float[] output, int width, int height)
        {
            var image = new Image<Rgb24>(width, height);

            image.ProcessPixelRows(accessor =>
            {
                for (int h = 0; h < height; h++)
                {
                    Span<Rgb24> pixelRow = accessor.GetRowSpan(h);
                    for (int w = 0; w < width; w++)
                    {
                        // Convert from CHW to HWC
                        int rIndex = 0 * height * width + h * width + w;
                        int gIndex = 1 * height * width + h * width + w;
                        int bIndex = 2 * height * width + h * width + w;

                        // CRITICAL: Output is already in [0, 255] range!
                        // DO NOT multiply by 255 again!
                        // Just clamp and convert to byte
                        byte r = (byte)Math.Clamp(output[rIndex], 0, 255);
                        byte g = (byte)Math.Clamp(output[gIndex], 0, 255);
                        byte b = (byte)Math.Clamp(output[bIndex], 0, 255);

                        pixelRow[w] = new Rgb24(r, g, b);
                    }
                }
            });

            return image;
        }

        /// <summary>
        /// Run LaMa ONNX inference
        /// </summary>
        public static void RunInference(string modelPath, string imagePath, string maskPath, string outputPath)
        {
            Console.WriteLine($"Loading ONNX model from {modelPath}...");
            using var session = new InferenceSession(modelPath);

            Console.WriteLine($"Loading image from {imagePath}...");
            using var originalImage = Image.Load<Rgb24>(imagePath);
            int originalWidth = originalImage.Width;
            int originalHeight = originalImage.Height;

            // Pad dimensions to be divisible by 8
            int paddedWidth = PadToModulo(originalWidth, 8);
            int paddedHeight = PadToModulo(originalHeight, 8);

            Console.WriteLine($"  Original size: {originalWidth}x{originalHeight}");
            Console.WriteLine($"  Padded size: {paddedWidth}x{paddedHeight}");

            // Resize to padded dimensions
            var image = originalImage.Clone(ctx => ctx.Resize(paddedWidth, paddedHeight));

            Console.WriteLine($"Loading mask from {maskPath}...");
            using var originalMask = Image.Load<L8>(maskPath);
            var mask = originalMask.Clone(ctx => ctx.Resize(paddedWidth, paddedHeight));

            // Preprocess inputs
            Console.WriteLine("Preprocessing inputs...");
            float[] imageArray = PreprocessImage(image);
            float[] maskArray = PreprocessMask(mask);

            // Create tensors with batch dimension
            var imageTensor = new DenseTensor<float>(imageArray, new[] { 1, 3, paddedHeight, paddedWidth });
            var maskTensor = new DenseTensor<float>(maskArray, new[] { 1, 1, paddedHeight, paddedWidth });

            Console.WriteLine($"Running inference...");
            Console.WriteLine($"  Image tensor shape: [{string.Join(", ", imageTensor.Dimensions)}]");
            Console.WriteLine($"  Mask tensor shape: [{string.Join(", ", maskTensor.Dimensions)}]");
            Console.WriteLine($"  Image range: [{imageArray.Min():F3}, {imageArray.Max():F3}]");
            Console.WriteLine($"  Mask range: [{maskArray.Min():F3}, {maskArray.Max():F3}]");

            // Run inference
            var inputs = new[]
            {
                NamedOnnxValue.CreateFromTensor("image", imageTensor),
                NamedOnnxValue.CreateFromTensor("mask", maskTensor)
            };

            using var results = session.Run(inputs);
            var output = results.First().AsEnumerable<float>().ToArray();

            Console.WriteLine($"  Output length: {output.Length}");
            Console.WriteLine($"  Output range: [{output.Min():F3}, {output.Max():F3}]");

            // Postprocess output
            Console.WriteLine("Postprocessing output...");
            using var resultImage = PostprocessOutput(output, paddedWidth, paddedHeight);

            // Crop to original size
            resultImage.Mutate(ctx => ctx.Crop(new Rectangle(0, 0, originalWidth, originalHeight)));

            // Save result
            Console.WriteLine($"Saving result to {outputPath}...");
            resultImage.Save(outputPath);

            Console.WriteLine("✅ Done!");
            Console.WriteLine("\nIMPORTANT NOTES:");
            Console.WriteLine("  - Input was normalized to [0, 1] range");
            Console.WriteLine("  - Output is already in [0, 255] range (NO additional scaling)");
            Console.WriteLine("  - Result saved as uint8 image");
        }

        public static void Main(string[] args)
        {
            if (args.Length != 4)
            {
                Console.Error.WriteLine("Usage: dotnet run <model.onnx> <input.jpg> <mask.png> <output.jpg>");
                Console.Error.WriteLine("");
                Console.Error.WriteLine("For more information, see the README's \"ONNX Preprocessing & Postprocessing Guide\" section.");
                Environment.Exit(1);
                return;
            }

            string modelPath = args[0];
            string imagePath = args[1];
            string maskPath = args[2];
            string outputPath = args[3];

            try
            {
                RunInference(modelPath, imagePath, maskPath, outputPath);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Console.Error.WriteLine(ex.StackTrace);
                Environment.Exit(1);
            }
        }
    }
}
